### ISSUE TITLE
Updating assets is broken in v13.0.0

### ISSUE THREAD

**Issue by ClaytonPassmore** [Shopify/shopify-api-ruby#1164]
# Issue summary

In v12.5.0, you could update the value of an asset by fetching it, setting a new value, and then saving it:

```ruby
asset = ShopifyAPI::Asset.all(asset: { key: "layout/theme.liquid" }, theme_id: theme.id).first
asset.value = asset.value + "<p>Something new</p>\n"
asset.save!
```

In v13.0.0, updating an asset results in a 406 Not Acceptable error.

I'm fairly confident this is broken because of the changes made in #1149. We're no longer sending up the required `key` parameter (because it hasn't changed).

With access to the asset API being removed in API version 2023-04, I'm not super hopeful that this is actually going to get fixed 🙈 

- `shopify_api` version: v13.0.0
- Ruby version: 3.2.2
- Operating system: Ubuntu 22.04 LTS

## Expected behavior

Updating an asset should behave similar to the way it did in v12.5.0 (in that it should work 😄). Including the `key` when we save the model would fix the problem.

v12.5.0 was obviously sending up more attributes than it should have, but it did work. See the screenshot below. Excuse the large code block - I'm replicating how the gem (incorrectly) determined which attributes changed in v12.5.0.

<img width="949" alt="Screenshot 2023-05-25 at 3 10 58 PM" src="https://github.com/Shopify/shopify-api-ruby/assets/6856391/72629344-fe88-4e11-ae16-c25acc0f127e">

## Actual behavior

In v13.0.0, updating an asset results in a 406 Not Acceptable error. Note that I'm using the new private methods introduced in #1149 for determining the changed attributes.

<img width="953" alt="Screenshot 2023-05-25 at 3 28 18 PM" src="https://github.com/Shopify/shopify-api-ruby/assets/6856391/9667c8f9-87f9-4030-a172-ac465a01d124">

## Steps to reproduce the problem

1. Fetch an asset
1. Change the asset's `value`
1. Try to save the asset

**Comment by garethson** [comment]
https://github.com/Shopify/shopify-api-ruby/pull/1149 also appeared to break saving metafields. Example:
```ruby
metafield = ShopifyAPI::Metafield.all(limit: 1, namespace: "foo", key: "bar").first
metafield.value = "new value"
metafield.save
```
This returns `{"errors":{"metafield":"Required parameter missing or invalid"}` an error because neither `key` nor `namespace` are sent, as those fields did not change.

**Comment by nickmealey** [comment]
I'm also seeing this issue. If the key/value aren't changed they aren't included in the request.

**Comment by nickmealey** [comment]
Looking at this closer it appears this commit is causing the issue: https://github.com/Shopify/shopify-api-ruby/commit/9a7ef8a6a22548b6ad7049ecb02b86f598b5f3a5

**Comment by flavio-b** [comment]
Yup, we're seeing the same issue. I've opened a ticket with Shopify via Partner Support and reference this thread.

**Comment by flavio-b** [comment]
While updating the value directly used to work, I noticed the [example in the official docs](https://shopify.dev/docs/api/admin-rest/2022-04/resources/asset#put-themes-theme-id-assets), about updating an Asset's value, actually involves creating a new instance, setting the key and value manually, and saving. That seems to do the tricky.

I'm not sure if the error we're seeing now is the expected behaviour.

**Comment by nelsonwittwer** [comment]
From what I can tell there are few problems going on with tracking attribute changes outstanding:

1. Required fields for some resources are not being sent in update requests
2. Attributes that were added after creation don't work

## Required Fields

This worked in previous versions because we were sending the entire state of the current object which led to other big unexpected issues as reported in https://github.com/Shopify/shopify-api-ruby/issues/1133 and https://github.com/Shopify/shopify-api-ruby/issues/1037. Since we are now only sending modified fields, this has the unfortunate side affect of not always sending all the required fields.

The solution to this issue would be to send the `attributes_to_update` as well as the required fields needed for each request. Unfortunately, we don't have the concept of required fields in our REST resources right now :( 

The OpenAPI schemas that generate these resources do know the required fields for each resource, however; so getting at these required fields in the resources should be doable. I will open an issue to look into adding that functionality within our next release.

### Work around
Reverting to sending the whole modified object isn't an option as it had it's own bad consequences as earlier noted. 

I'd recommend using using @flavio-b 's [suggested work around for the Asset resource](https://github.com/Shopify/shopify-api-ruby/issues/1164#issuecomment-1586292385) or using the GraphQL interface for modifying assets or any other similarly affected REST resource until we can add those required fields in our next release.

## Attributes added after creation
This will be fixed with #1197

**Comment by luklapp** [comment]
@nelsonwittwer I have a similar issue when updating arrays. Example:

```
@order = ShopifyAPI::Order.find(id: params[:id])
@order.note_attributes = @order.note_attributes + [NEW ATTRIBUTES]
@order.save!
```

It produces the broken diff mentioned by @mvanholstyn. My current workaround looks like this:

```
@order.note_attributes = @order.note_attributes + [{'name' => 'foo', 'value' => 'bar'}]
@order.original_state['note_attributes'] = nil

@order.save!
```

Resetting the `note_attributes` of the `original_state` to nil produces the correct diff and saving works.

**Comment by mvanholstyn** [comment]
@luklapp what I found works for me is instead of finding the order, doing this..

```
@order = ShopifyAPI::Order.new
@order.id = params[:id]
@order.note_attributes = @order.note_attributes + [NEW ATTRIBUTES]
@order.save!
```

Doing it this way causes it to only update the attributes you set, and the attributes you don't set get left as-is.
